swarm/network/discovery_test.go

```diff
@@ -58,3 +65,137 @@ func TestDiscovery(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// TestSubpeersMsg tests that the correct set of peers is suggested
+// to another peer with a given depth and kademlia
+func TestSubpeersMsg(t *testing.T) {
+
+	// This is the defined depth
+	testDepth := 2
+
+	// construct ProtocolTester and hive
+	params := NewHiveParams()
+	s, hive, err := newHiveTester(t, params, 1, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// register some addresses in specific bins (must coincide with testDepth)
+	registerBzzAddr(0, hive, true)  // bin 0
+	registerBzzAddr(0, hive, true)  // bin 0
+	registerBzzAddr(1, hive, true)  // bin 1
+	registerBzzAddr(1, hive, true)  // bin 1
+	registerBzzAddr(3, hive, true)  // bin 3
+	registerBzzAddr(4, hive, true)  // bin 4
+	registerBzzAddr(3, hive, false) // add a known but not connected peer
+	registerBzzAddr(1, hive, false) // add a known but not connected peer
+
+	// start the hive
+	hive.Start(s.Server)
+	defer hive.Stop()
+
+	// the pivot node is the only one from the ProtocolTester
+	pivot := s.Nodes[0]
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// we need to wait until the pivot node is actually connected to our hive
+WAIT_PIVOT:
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for the pivot node to connect")
+		case <-time.After(100 * time.Millisecond):
+			if _, ok := hive.peers[pivot.ID()]; ok {
+				break WAIT_PIVOT
+			}
+		}
+	}
+
+	// get BzzAddr of the pivot
+	pivotAddress := hive.peers[pivot.ID()]
+	pivotBzz := pivotAddress.BzzAddr.Over()
+
+	// now we need to identify which peers are expected
+	// iterate the hive's connection and only add peers below testDepth
+	var expectedPeers []*BzzAddr
+	hive.EachConn(hive.BaseAddr(), 256, func(p *Peer, po int) bool {
```

Comment:
again, you will need the  `pivotAddress` not self, and 255 as you dont want to send the peer to itself